### DISCUSSION TITLE
Compatibility with Next.js v13 React Server Components

### DISCUSSION THREAD

**Discussion by Belco90** [garmeeh/next-seo#1068] [Q&A] [answered]
Hi there!

I'm a huge fan of `next-seo`, it really simplifies the whole SEO wire-up in my Next.js projects.

I'm migrating a simple project to [the new `app/` architecture with Next.js v13](https://beta.nextjs.org/docs/upgrade-guide#migrating-from-pages-to-app), and I'm afraid `next-seo` doesn't work anymore.

It kind of makes sense: Next.js complains it can't render complex elements from the server, with an error like this:

![image](https://user-images.githubusercontent.com/2677072/198559484-f7bfd17a-0709-468c-946e-6ee584ebbc2d.png)

I realized this is caused by `NextSeo` or `DefaultSeo` used in React Server Component (which are all components by default in Next.js v13 using the App directory).

I moved my pages using `next-seo` to Client Component. Now, the project works fine, but there are no SEO tags included. It's too late at this point for injecting any meta tag, right? So I think this is expected.

I can put the meta tags myself in my Server pages components, for sure, but I liked how `next-seo` simplified this process. I was basically wondering if there is any plan to make this package compatible with the RSC, or we are expected to just set the meta tags manually.

**Comment by Belco90** [comment]
I just found you can [modify the `<head>` tag using a new special `head.js` component](https://beta.nextjs.org/docs/routing/pages-and-layouts#modifying-head). This is what I'll use for now, but it makes me wonder if `next-seo` makes any sense in all this then.

**Reply by garmeeh** [reply]
Hey @Belco90 👋 

I need to modify NextSeo a little and write up some documentation for Next.js v13. It will work slightly different now that we have `head.js` components. 

Hoping to have something out mid next week.

**Reply by Belco90** [reply]
Amazing! Thanks for your quick answer. Looking forward to the new docs.

**Reply by eric-mathison** [reply]
Just keep in mind there warning:
Warning: Currently, the <Head> export does not re-render on client-side transition using next/link, only on initial render and reloads. To work around this for <title>, you can use a client component with useEffect that updates document.title. We plan to fix this in a future release.

**Reply by lveillard** [reply]
Thanks @garmeeh ! Happy to test it once is out ;)

**Reply by dejurin** [reply]
head.tsx does not provide such advanced features as next-seo.

**Reply by garmeeh** [reply]
Hey @eric-mathison do you have link to the issue on that warning by any chance? Would be good to up to date with it.

I don't think I will add any workaround in `next-seo` for this as it is an upstream issue that is going to be fixed.

Thank you for highlighting this! I would probably have lost hours pulling my hair out wondering why it wasn't working locally 😅

**Reply by eric-mathison** [reply]
I found it while reading up on the docs here: https://beta.nextjs.org/docs/routing/pages-and-layouts#modifying-head
I wasn't able to find any issue discussing it more though. I wish there was some idea of when in 'the future' they are planning on tackling this. 
I realize that this doesn't necessarily affect google or any crawlers individually loading URLs, but I don't particularly want to create a client component for every page either just to change the meta in the head. 
I'm uncertain what all the side effects of this will be though and what information will need to be put in the client component (useeffect). Will any social share buttons pull the wrong og data for instance? I'll have to play around and see. It might not be as big of a deal and for visual sake, the title might be the only thing needing to be updated.

**Comment by garmeeh** [comment]
I spent the day yesterday playing around with different ways to adapt how NextSeo will work with the new `app` directory. 

I also spent some time researching how the new flow works with `head.js` and `layout.js`. If anyone would like to check that out, you can find my quick summary here on the [feedback for the Next.js team](https://github.com/vercel/next.js/discussions/41745#discussioncomment-4011958). I have asked them for clarification on what the intentional behaviour is. This will steer the direction of how I adapt `next-seo`.

I have an idea of how to approach it so once I get clarification I will refactor what I have and get it live for testing.

**Reply by Belco90** [reply]
@garmeeh amazing! Thanks for the quick feedback.

**Comment by SimonVillage** [comment]
So just to confirm, currently there is no way to migrate to app folder structure as long as `next-seo` isn't compatible?

**Comment by garmeeh** [comment, marked as answer]
🚢  `next-seo@5.13.0`  now allows you to opt in to app directory. Please make sure to read the [docs](https://github.com/garmeeh/next-seo/blob/master/APP_DIRECTORY.md) I wrote detailing a little bit about how things of changed. 

Open to feedback, suggestions and info on potential bugs [here](https://github.com/garmeeh/next-seo/discussions/1073)

**Reply by Belco90** [reply]
Mant thanks! I'll give it a try during this week and leave feedback in the provided link.
